File: Java/Algoritmos/src/MinPriorityQueue.java
import java.util.Arrays;

public class MinPriorityQueue {

  private int capacity;
  private int size;
  private Integer[] array;
  
  public MinPriorityQueue(int capacity) {
    this.capacity = capacity;
    this.array = new Integer[capacity];
    this.size = 0;
  }
  
  public void add(int i) {
    if (this.size == 0 || this.size < this.capacity || i > this.array[0]) {
      if (this.size == this.capacity) {
        this.poll();
      }
      this.array[this.size] = i;
      this.swim(this.size++);
    }
  }
  
  public int peek() {
    return this.array[0];
  }

  public int poll() {
    this.exchange(0, --this.size);
    int aux = this.array[this.size];
    this.array[this.size] = null;
    this.sink(0);
    return aux;
  }
  
  public void exchange(int i, int j) {
    int aux = this.array[i];
    this.array[i] = this.array[j];
    this.array[j] = aux;
  }
  
  public void swim(int i) {
    while (i > 0 && ((i - 1) / 2) >= 0) {
      int parent = (i - 1) / 2;
      if (this.array[i] < this.array[parent]) {
         this.exchange(i, parent);
      }
      i = parent;
    }
  }
  
  public void sink(int i) {
    while (i < this.size && ((i * 2) + 1) < this.size) {
       int leftChild = (i * 2) + 1;
       int rightChild = (i * 2) + 2;
       if (rightChild < this.size && 
           this.array[rightChild] < this.array[leftChild]) {
         leftChild = rightChild;
       }
       if (this.array[i] > this.array[leftChild]) {
         this.exchange(i, leftChild);
       }
       i = leftChild;
    }
  }

  @Override
  public String toString() {
    return Arrays.toString(this.array);
  }

  public static void main(String[] args) {
      MinPriorityQueue pq = new MinPriorityQueue(5);
      
//      pq.add(11);
//      pq.add(10);
//      pq.add(2);
//      pq.add(9);
//      pq.add(3);
//      pq.add(8);
//      pq.add(4);
//      pq.add(7);
//      pq.add(5);
//      pq.add(6);
      
//      pq.add(1);
//      pq.add(2);
//      pq.add(3);
//      pq.add(4);
//      pq.add(5);
//      pq.add(6);
//      pq.add(7);
//      pq.add(8);
//      pq.add(9);
//      pq.add(10);
      
      pq.add(10);
      pq.add(9);
      pq.add(8);
      pq.add(7);
      pq.add(6);
      pq.add(5);
      pq.add(4);
      pq.add(3);
      pq.add(2);
      pq.add(1);
      pq.add(0);

      System.out.println(pq); 
  }

}

File: Python/MinPriorityQueue.py
class MinPriorityQueue:

    def __init__(self, capacity):
        self.capacity = capacity
        self.array = []
        self.size = 0

    def add(self, i):
        self.array.append(i)
        self.swim(self.size)
        self.size += 1
        if(self.size > self.capacity):
            self.poll()

    def peek(self):
        return self.array[0]

    def poll(self):
        self.exchange(0, self.size - 1)
        self.sink(0)
        self.size -= 1
        return self.array.pop()

    def exchange(self, i, j):
        aux = self.array[i]
        self.array[i] = self.array[j]
        self.array[j] = aux

    def swim(self, i):
        while i > 0 and ((i - 1) / 2) >= 0:
            parent = (i - 1) / 2
            if self.array[i] < self.array[parent]:
                self.exchange(i, parent)
            i = parent

    def sink(self, i):
        while i < self.size and ((i * 2) + 1) < self.size:
            left_child = (i * 2) + 1
            right_child = (i * 2) + 2
            if (right_child < self.size and
                    self.array[right_child] > self.array[left_child]):
                left_child = right_child
            if self.array[i] > self.array[left_child]:
                self.exchange(i, left_child)
            i = left_child

    def __str__(self):
        return "capacity: " + str(self.capacity) + \
                " elements: " + str(self.array)

if __name__ == "__main__":
    pq = MinPriorityQueue(5)
    print pq
    pq.add(11)
    print pq
    pq.add(10)
    print pq
    pq.add(2)
    print pq
    pq.add(9)
    print pq
    pq.add(3)
    print pq
    pq.add(8)
    print pq
    pq.add(4)
    print pq
    pq.add(7)
    print pq
    pq.add(5)
    print pq
    pq.add(6)

    print pq